### PR TITLE
[pipelinable]use pipelinable context to initialize non-pipeline model

### PR DESCRIPTION
**Pull request by YuliangLiu0306**

### CODE REVIEW

File: colossalai/nn/layer/layer_spec.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+    @property
+    def param_count(self):
+        return self._param_count
+
+    def build(self):
+        """Build the stored specification."""
+        return self.typename(*self.module_args, **self.module_kwargs)
+
+    def get_children(self, children):
```
Comment (FrankLeeeee) [inline colossalai/nn/layer/layer_spec.py:124]: Should it be `set_children`?

Comment (YuliangLiu0306) [reply colossalai/nn/layer/layer_spec.py:124]: yes

**Comment by kurisusnowdeng** [comment]
What's the purpose of this layer spec?

**Comment by feifeibear** [comment]
I suppose it is mainly used to count the param numel of each layer, and used for optimization of PP.

**Comment by YuliangLiu0306** [comment]
> What's the purpose of this layer spec?

It's a module lazy init feature. Model built under this context will not build real instance but a list of layer_spec.

**Comment by kurisusnowdeng** [comment]
> > What's the purpose of this layer spec?
> 
> It's a module lazy init feature. Model built under this context will not build real instance but a list of layer_spec. Then we will use those spec to find a partition policy. Finally, the partitioned model instance will be built on their own rank.

Great. Is it for auto-parallelization? I look forward very much to a overview of the functionality when it is completed.

File: colossalai/utils/model/pipelinable.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+    @property
+    def layers_count(self):
+        return len(self._layer_spec_list)
+
+    @property
+    def funcs_count(self):
+        return len(self._func_dict)
+
+    def __enter__(self):
```
Comment (feifeibear) [inline colossalai/utils/model/pipelinable.py:31]: It has been implemented in InsertPostInitMethodToModuleSubClasses. Why override?

File: colossalai/utils/model/pipelinable.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+            raise RuntimeError('LayerSpec only supports torch.nn.Module types.')
+
+    def __repr__(self):
+        return call_to_str(self.typename.__name__, self.module_args, self.module_kwargs)
+
+    @property
+    def param_count(self):
+        return self._param_count
+
+    def build(self):
```
Comment (feifeibear) [inline colossalai/utils/model/pipelinable.py:222]: can you handle the case where a param sharded by two submodules?

File: colossalai/utils/model/pipelinable.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+            modified_args.append(obj)
+        # (lyl)TODO: analyse kwargs as well
+        modified_args = tuple(modified_args)
+        self._root_children = list(module.children())
+        self._model = module
+        layer_spec = LayerSpec(module.__class__, *modified_args, **kwargs)
+        layer_spec.set_children(module.children())
+        self._layer_spec_dict[module_id] = layer_spec
+        for param in module.parameters(recurse=False):
+            param.data = torch.rand(1, 1)
```
[the remark this answers is not in the document]
Comment (YuliangLiu0306) [reply colossalai/utils/model/pipelinable.py:101]: I may use ColoTensor to replace this dummy tensor. An unexpected bug occured when I use torch.empty() during testing, it requires 2 dimension tensor. Bug has been fixed, but I would like to use ColoTensor to replace this dummy tensor if the ColoTensor lazy init feature works as I expected.

File: colossalai/utils/model/pipelinable.py (hunk 3 above)
Comment (YuliangLiu0306) [reply colossalai/utils/model/pipelinable.py:222]: If the "build" process happens under the ZeroInit context, params will be sharded correctly.

File: colossalai/utils/model/pipelinable.py (hunk 2 above)
Comment (YuliangLiu0306) [reply colossalai/utils/model/pipelinable.py:31]: There is a tiny difference in the _post_init_method argument list. In the base class, _post_init_method only need module, but pipelinable need args and key world args to create the layer spec.

File: colossalai/utils/model/pipelinable.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
+import torch
+import functools
+from colossalai.utils.model.utils import _substitute_init_recursively, InsertPostInitMethodToModuleSubClasses, call_to_str
+from colossalai.builder.pipeline import partition_uniform, partition_balanced
+from colossalai.core import global_context as gpc
+
+
+class Pipelinable(InsertPostInitMethodToModuleSubClasses):
```
[the remark this answers is not in the document]
Comment (YuliangLiu0306) [reply colossalai/utils/model/pipelinable.py:8]: Fine.

File: colossalai/utils/model/pipelinable.py (hunk 2 above)
Comment (feifeibear) [reply colossalai/utils/model/pipelinable.py:31]: add  args and key to InsertPostInitMethodToModuleSubClasses?

File: colossalai/utils/model/pipelinable.py (hunk 3 above)
Comment (feifeibear) [reply colossalai/utils/model/pipelinable.py:222]: Sorry, I mean shared (not sharded) by two submodules.

File: colossalai/utils/model/pipelinable.py (hunk 2 above)
Comment (YuliangLiu0306) [reply colossalai/utils/model/pipelinable.py]: Done.

**Review by feifeibear — APPROVED** [review]

**Pull request merged**